integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -776,33 +777,84 @@ private int chownInternal(String path, long uid, long gid) {
   }
 
   /**
-   * Truncate is not supported internally by Alluxio.
-   * Truncate is supported here only for a special overwrite case.
-   * Libfuse issues open() - truncate() to size 0 - write() new contents - release()
-   * to overwrite an existing file. Since files can be written only once,
-   * only sequentially, and never be modified in Alluxio, we delete the existing file
-   * and create a new file for writing in open() and consider truncate() to size 0 as
-   * a noop to fulfill the overwrite requirement.
+   * Since files can be written only once, only sequentially,
+   * and never be modified in Alluxio, truncate is not supported internally by Alluxio.
+   * 
+   * In Alluxio Fuse, we support truncate in some special cases.
    *
    * @param path the file to truncate
    * @param size the size to truncate to
    * @return 0 if succeed, error code otherwise
    */
   @Override
   public int truncate(String path, long size) {
-    LOG.debug("truncate {} to {}", path, size);
+    return AlluxioFuseUtils.call(LOG, () -> truncateInternal(path, size),
+        "Fuse.Truncate", "path=%s,size=%d", path, size);
+  }
+
+  private int truncateInternal(String path, long size) {
+    // Truncate cases supported:
+    // 1. Truncate size = 0, file does not exist => no-op
+    // 2. Truncate size = 0, file exists => delete file, remove create file entries if any
+    // 3. Truncate size != 0, file does not exist => error out
+    // 4. Truncate size != 0, file is completed 
+    // => no-op if file size = truncate size, error out otherwise
+    // 5. Truncate size != 0, file is being written by other applications 
+    // => error out, don't know exact file written size
+    // 6. Truncate size != 0, file is being written by current Fuse 
+    // => no-op if written size = truncate size, error out otherwise
+    final AlluxioURI uri = mPathResolverCache.getUnchecked(path);
+    URIStatus status = null;
+    try {
+      status = mFileSystem.getStatus(uri);
+    } catch (FileDoesNotExistException | InvalidPathException e) {
+      status = null;
+    } catch (Throwable t) {
+      LOG.error("Failed to truncate path {} to {}. Failed to get file status", path, size, t);
+      return -ErrorCodes.EIO();
+    }
+    CreateFileEntry<FileOutStream> ce = mCreateFileEntries.getFirstByField(PATH_INDEX, path);
     if (size == 0) {
-      // truncate may be called in overwrite process:
```

Comment:
Is it safe if we delete the file which is written by others?

Reply:
Good catch!